arangod/RocksDBEngine/RocksDBIndexCacheRefillFeature.cpp

```diff
@@ -0,0 +1,341 @@
+  startsAfter<RocksDBEngine>();
+
+  // default value must be at least 1, as the minimum allowed value is also 1.
+  TRI_ASSERT(_maxConcurrentIndexFillTasks >= 1);
+}
+
+RocksDBIndexCacheRefillFeature::~RocksDBIndexCacheRefillFeature() {
+  stopThread();
+}
+
+void RocksDBIndexCacheRefillFeature::collectOptions(
+    std::shared_ptr<options::ProgramOptions> options) {
+  options
+      ->addOption(
+          "--rocksdb.auto-fill-index-caches-on-startup",
+          "Automatically fill in-memory index cache entries on server startup.",
+          new options::BooleanParameter(&_fillOnStartup),
+          arangodb::options::makeFlags(
+              options::Flags::DefaultNoComponents, options::Flags::OnDBServer,
+              options::Flags::OnSingle, options::Flags::Hidden,
+              options::Flags::Experimental))
+      .setIntroducedIn(30906)
+      .setIntroducedIn(31020);
+
+  options
+      ->addOption("--rocksdb.auto-refill-index-caches-on-modify",
+                  "Automatically (re-)fill in-memory index cache entries upon "
+                  "insert/update/replace.",
+                  new options::BooleanParameter(&_autoRefill),
+                  arangodb::options::makeFlags(
+                      options::Flags::DefaultNoComponents,
+                      options::Flags::OnDBServer, options::Flags::OnSingle,
+                      options::Flags::Hidden, options::Flags::Experimental))
+      .setIntroducedIn(30906)
+      .setIntroducedIn(31020);
+
+  options
+      ->addOption(
+          "--rocksdb.auto-refill-index-caches-queue-capacity",
+          "Maximum capacity for automatic in-memory index cache refill queue.",
+          new options::UInt64Parameter(&_maxCapacity),
+          options::makeFlags(options::Flags::DefaultNoComponents,
+                             options::Flags::OnDBServer,
+                             options::Flags::OnSingle, options::Flags::Hidden,
+                             options::Flags::Experimental))
+      .setIntroducedIn(30906)
+      .setIntroducedIn(31020);
+
+  options
+      ->addOption("--rocksdb.max-concurrent-index-fill-tasks",
+                  "Maximum number of concurrent index fill tasks at startup.",
+                  new options::UInt64Parameter(&_maxConcurrentIndexFillTasks),
+                  options::makeFlags(
+                      options::Flags::DefaultNoComponents,
+                      options::Flags::OnDBServer, options::Flags::OnSingle,
+                      options::Flags::Hidden, options::Flags::Experimental))
+      .setIntroducedIn(30906)
+      .setIntroducedIn(31020);
+}
+
+void RocksDBIndexCacheRefillFeature::validateOptions(
+    std::shared_ptr<options::ProgramOptions> options) {
+  if (_maxConcurrentIndexFillTasks == 0) {
+    // number of tasks cannot be 0
+    _maxConcurrentIndexFillTasks = 1;
+  }
+}
+
+void RocksDBIndexCacheRefillFeature::beginShutdown() {
+  if (_refillThread != nullptr) {
+    _refillThread->shutdown();
```

Comment:
`shutdown` blocks until the thread has finished. I suppose we want to call `beginShutdown` here instead.
```suggestion
    _refillThread->beginShutdown();
```